modules/replicator/src/main/java/org/apache/ignite/internal/replicator/command/SafeTimePropagatingCommand.java

```diff
@@ -26,7 +27,11 @@
  * Common interface for commands carrying safe time.
  */
 public interface SafeTimePropagatingCommand extends WriteCommand {
-    /** Safe time. */
+    @Override
+    HybridTimestamp initiatorTime();
```

Comment:
Mmmm, I didn't get the concept.

- Initiator time is the time on the node that creates and sends the command, right? In case of partitions it's primary replica?
- On leader you do calculate safeTime as `safeTs = command.initiatorTime() == null ? clock.now() : clock.now(command.initiatorTime())` and then patch the command with the safeTime. 

All in all that means that Initiator time is a sort of timestampAware messaging between the node that hosts raft-client and raft leader and safeTime mechanics is now fully managed inside raft. Is that correct?

Reply:
Yes, safe time mechanics is fully managed inside raft.
initiatorTime is required to establish ordering between primary replica and replication group.
For example, for full transaction safe timestamp (which is equal to commit timestamp) must be greater than start timestamp.
Metastorage already has the same concept: MetaStorageWriteCommand#initiatorTime